Never transfer ownership by a raw pointer

In LinearAnimationBuilder the operator `LinearAnimation<T>*` returns a heap allocated object which isn't deleted hence causing a memory leak.

raw pointers donot represent ownership and by giving away the ownership there is a leak.
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Ri-raw

use `std::unique_ptr` instead.

weird, i checked in clang-tidy 

```
┌─[voldyman@arc:~/Code/beatwave/build on master]
└─> /usr/local/Cellar/llvm/3.8.1/bin/clang-tidy  ../src/**/*.cpp
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
/Users/voldyman/Code/beatwave/src/beatwave/game.cpp:55:29: warning: Potential memory leak [clang-analyzer-cplusplus.NewDeleteLeaks]
    dummy.animate(repeat(2, seq<sf::Vector2f>({
                            ^
/Users/voldyman/Code/beatwave/src/beatwave/game.cpp:56:9: note: Calling 'LinearTransitionBuilder::operator LinearTransition<sf::Vector2<float> > *'
        from(sf::Vector2f(100.0f, 100.0f)).to(sf::Vector2f(200.0f, 200.0f)).during(1000),
        ^
/Users/voldyman/Code/beatwave/src/core/lineartransitionbuilder.hpp:33:16: note: Memory is allocated
        return new LinearTransition<T>(m_initialState, m_transitionTime, m_finalState);
               ^
/Users/voldyman/Code/beatwave/src/beatwave/game.cpp:56:9: note: Returned allocated memory
        from(sf::Vector2f(100.0f, 100.0f)).to(sf::Vector2f(200.0f, 200.0f)).during(1000),
        ^
/Users/voldyman/Code/beatwave/src/beatwave/game.cpp:55:29: note: Potential memory leak
    dummy.animate(repeat(2, seq<sf::Vector2f>({
                            ^
```

it's definitely a memory leak, since all the object is being allocated but never deleted, i am not sure why valgrind doesn't show that.

--- a/src/beatwave/game.cpp
+++ b/src/beatwave/game.cpp
@@ -5,7 +5,7 @@
 #include <beatwave/game.hpp>
 #include <core/util.hpp>
 #include <core/lineartransitionbuilder.hpp>
-#include <core/seqcombinator.hpp>
+#include <core/seqcombinatorbuilder.hpp>
 #include <core/forever.hpp>
 #include <core/repeat.hpp>
 
@@ -52,11 +52,10 @@ Game::Game():
            sf::Color::White),
     dummy(sf::Vector2f(100.0f, 100.0f))
 {
-    dummy.animate(repeat(2, seq<sf::Vector2f>({
-        from(sf::Vector2f(100.0f, 100.0f)).to(sf::Vector2f(200.0f, 200.0f)).during(1000),
-        from(sf::Vector2f(200.0f, 200.0f)).to(sf::Vector2f(200.0f, 100.0f)).during(100),
-        from(sf::Vector2f(200.0f, 100.0f)).to(sf::Vector2f(100.0f, 100.0f)).during(1000)
-    })));
+    dummy.animate(repeat<sf::Vector2f>(2,
+                         start<sf::Vector2f>(from(sf::Vector2f(100.0f, 100.0f)).to(sf::Vector2f(200.0f, 200.0f)).during(1000))
+                         .then(from(sf::Vector2f(200.0f, 200.0f)).to(sf::Vector2f(200.0f, 100.0f)).during(100))
+                         .then(from(sf::Vector2f(200.0f, 100.0f)).to(sf::Vector2f(100.0f, 100.0f)).during(1000))));
 }
 
 

--- a/src/core/animated.hpp
+++ b/src/core/animated.hpp
@@ -25,9 +25,9 @@ public:
         m_animation.reset();
     }
 
-    void animate(Animation<T> *animation)
+    void animate(AnimationPtr<T> &&animation)
     {
-        m_animation.reset(animation);
+        m_animation = std::move(animation);
 
         if (!m_animation->isFinished()) {
             m_value = m_animation->getCurrentState();
@@ -43,7 +43,7 @@ public:
 
 private:
     T m_value;
-    std::unique_ptr<Animation<T>> m_animation;
+    AnimationPtr<T> m_animation;
 };
 
 #endif  // ANIMATED_HPP_

--- a/src/core/animation.hpp
+++ b/src/core/animation.hpp
@@ -14,4 +14,7 @@ public:
     virtual void reset() = 0;
 };
 
+template <typename State>
+using AnimationPtr = std::unique_ptr<Animation<State>>;
+
 #endif  // ANIMATION_HPP_

--- a/src/core/forever.hpp
+++ b/src/core/forever.hpp
@@ -8,8 +8,8 @@ template <typename State>
 class Forever: public Animation<State>
 {
 public:
-    Forever(Animation<State> *animation):
-        m_animation(animation)
+    Forever(AnimationPtr<State> &&animation):
+        m_animation(std::move(animation))
     {}
 
     State nextState(const sf::Int32 deltaTime) override
@@ -35,14 +35,14 @@ public:
     {}
 
 private:
-    std::unique_ptr<Animation<State>> m_animation;
+    AnimationPtr<State> m_animation;
 };
 
 
 template <typename State>
-Forever<State> *forever(Animation<State> *animation)
+AnimationPtr<State> forever(AnimationPtr<State> &&animation)
 {
-    return new Forever<State>(animation);
+    return AnimationPtr<State>(new Forever<State>(std::move(animation)));
 }
 
 #endif  // FOREVER_HPP_

--- a/src/core/lineartransitionbuilder.hpp
+++ b/src/core/lineartransitionbuilder.hpp
@@ -29,8 +29,8 @@ public:
         return *this;
     }
 
-    operator LinearTransition<T>*() {
-        return new LinearTransition<T>(m_initialState, m_transitionTime, m_finalState);
+    operator AnimationPtr<T>() {
+        return AnimationPtr<T>(new LinearTransition<T>(m_initialState, m_transitionTime, m_finalState));
     }
 
 private:
@@ -38,7 +38,6 @@ private:
     T m_finalState;
     sf::Int32 m_transitionTime;
 };
-
 
 template <typename T>
 LinearTransitionBuilder<T> from(const T &initialState)

--- a/src/core/repeat.hpp
+++ b/src/core/repeat.hpp
@@ -10,9 +10,9 @@ template <typename State>
 class Repeat: public Animation<State>
 {
 public:
-    Repeat(int counter, Animation<State> *animation):
+    Repeat(int counter, AnimationPtr<State> &&animation):
         m_initialCounter(std::max(0, counter)),
-        m_animation(animation),
+        m_animation(std::move(animation)),
         m_currentCounter(m_initialCounter)
     {}
 
@@ -44,14 +44,14 @@ public:
 
 private:
     const int m_initialCounter;
-    std::unique_ptr<Animation<State>> m_animation;
+    AnimationPtr<State> m_animation;
     int m_currentCounter;
 };
 
 template <typename State>
-Repeat<State> *repeat(int counter, Animation<State> *animation)
+AnimationPtr<State> repeat(int counter, AnimationPtr<State> &&animation)
 {
-    return new Repeat<State>(counter, animation);
+    return AnimationPtr<State>(new Repeat<State>(counter, std::move(animation)));
 }
 
 #endif  // REPEAT_HPP_

--- a/src/core/seqcombinator.hpp
+++ b/src/core/seqcombinator.hpp
@@ -10,13 +10,10 @@ template <typename State>
 class SeqCombinator: public Animation<State>
 {
 public:
-    SeqCombinator(std::initializer_list<Animation<State>*> animations):
+    SeqCombinator(std::vector<AnimationPtr<State>> &&animations):
+        m_animations(std::move(animations)),
         m_currentAnimation(0)
-    {
-        for (auto animation: animations) {
-            m_animations.push_back(std::unique_ptr<Animation<State>>(animation));
-        }
-    }
+    {}
 
     State nextState(const sf::Int32 deltaTime)
     {
@@ -57,11 +54,5 @@ private:
     std::vector<std::unique_ptr<Animation<State>>> m_animations;
     size_t m_currentAnimation;
 };
-
-template <typename State>
-SeqCombinator<State> *seq(std::initializer_list<Animation<State>*> animations)
-{
-    return new SeqCombinator<State>(animations);
-}
 
 #endif  // SEQCOMBINATOR_HPP_

--- a/src/core/seqcombinator.hpp
+++ b/src/core/seqcombinator.hpp
@@ -51,7 +51,7 @@ public:
     }
 
 private:
-    std::vector<std::unique_ptr<Animation<State>>> m_animations;
+    std::vector<AnimationPtr<State>> m_animations;
     size_t m_currentAnimation;
 };
 

--- a/src/core/seqcombinatorbuilder.hpp
+++ b/src/core/seqcombinatorbuilder.hpp
@@ -1,0 +1,36 @@
+#ifndef SEQCOMBINATORBUILDER_HPP_
+#define SEQCOMBINATORBUILDER_HPP_
+
+#include <core/seqcombinator.hpp>
+
+template <typename State>
+class SeqCombinatorBuilder
+{
+public:
+    SeqCombinatorBuilder(Animation<State> *animation)
+    {
+        m_animations.push_back(AnimationPtr<State>(animation));
+    }
+
+    SeqCombinatorBuilder<State> &then(Animation<State> *animation)
+    {
+        m_animations.push_back(AnimationPtr<State>(animation));
+        return *this;
+    }
+
+    operator Animation<State>*()
+    {
+        return new SeqCombinator<State>(std::move(m_animations));
+    }
+
+private:
+    std::vector<AnimationPtr<State>> m_animations;
+};
+
+template <typename State>
+SeqCombinatorBuilder<State> start(Animation<State> *animation)
+{
+    return SeqCombinatorBuilder<State>(animation);
+}
+
+#endif  // SEQCOMBINATORBUILDER_HPP_

--- a/src/core/seqcombinatorbuilder.hpp
+++ b/src/core/seqcombinatorbuilder.hpp
@@ -7,20 +7,20 @@ template <typename State>
 class SeqCombinatorBuilder
 {
 public:
-    SeqCombinatorBuilder(Animation<State> *animation)
+    SeqCombinatorBuilder(AnimationPtr<State> &&animation)
     {
-        m_animations.push_back(AnimationPtr<State>(animation));
+        m_animations.push_back(std::move(animation));
     }
 
-    SeqCombinatorBuilder<State> &then(Animation<State> *animation)
+    SeqCombinatorBuilder<State> &then(AnimationPtr<State> &&animation)
     {
-        m_animations.push_back(AnimationPtr<State>(animation));
+        m_animations.push_back(std::move(animation));
         return *this;
     }
 
-    operator Animation<State>*()
+    operator AnimationPtr<State>()
     {
-        return new SeqCombinator<State>(std::move(m_animations));
+        return AnimationPtr<State>(new SeqCombinator<State>(std::move(m_animations)));
     }
 
 private:
@@ -28,9 +28,9 @@ private:
 };
 
 template <typename State>
-SeqCombinatorBuilder<State> start(Animation<State> *animation)
+SeqCombinatorBuilder<State> start(AnimationPtr<State> &&animation)
 {
-    return SeqCombinatorBuilder<State>(animation);
+    return SeqCombinatorBuilder<State>(std::move(animation));
 }
 
 #endif  // SEQCOMBINATORBUILDER_HPP_

